File: ios/JustSpent/JustSpent/Common/Utilities/NumberPhraseParser.swift
//
//  NumberPhraseParser.swift
//  JustSpent
//
//  Comprehensive number phrase parser for voice commands
//  Supports basic numbers, hundreds, thousands, lakhs, crores, millions, billions, trillions, and decimals
//

import Foundation

/// Comprehensive number phrase parser for converting spoken numbers to numeric values
class NumberPhraseParser {

    // MARK: - Singleton
    static let shared = NumberPhraseParser()

    private init() {}

    // MARK: - Number Dictionaries

    private let basicNumbers: [String: Double] = [
        "zero": 0, "one": 1, "two": 2, "three": 3, "four": 4,
        "five": 5, "six": 6, "seven": 7, "eight": 8, "nine": 9,
        "ten": 10, "eleven": 11, "twelve": 12, "thirteen": 13, "fourteen": 14,
        "fifteen": 15, "sixteen": 16, "seventeen": 17, "eighteen": 18, "nineteen": 19
    ]

    private let tensNumbers: [String: Double] = [
        "twenty": 20, "thirty": 30, "forty": 40, "fifty": 50,
        "sixty": 60, "seventy": 70, "eighty": 80, "ninety": 90
    ]

    private let scaleNumbers: [String: Double] = [
        "hundred": 100,
        "thousand": 1_000,
        "lakh": 100_000, "lac": 100_000, "lakhs": 100_000, "lacs": 100_000,
        "million": 1_000_000,
        "crore": 10_000_000, "crores": 10_000_000,
        "billion": 1_000_000_000,
        "trillion": 1_000_000_000_000
    ]

    // MARK: - Public Methods

    /// Parse a number phrase and return the numeric value
    /// - Parameter text: The text containing a number phrase
    /// - Returns: The numeric value as a Double, or nil if parsing fails
    func parse(_ text: String) -> Double? {
        let normalizedText = normalizeText(text)

        // Try numeric extraction first (fast path)
        if let numericAmount = extractNumericAmount(normalizedText) {
            return numericAmount
        }

        // Fall back to written number parsing
        return parseWrittenNumber(normalizedText)
    }

    /// Extract amount from a voice command string
    /// - Parameter command: The voice command containing an amount
    /// - Returns: The extracted amount as Double, or nil if not found
    func extractAmountFromCommand(_ command: String) -> Double? {
        let normalizedCommand = normalizeText(command)

        // Patterns to extract amount substring from common voice command structures
        let amountPatterns = [
            #"(?:spent|spend|paid|pay|cost)\s+(.*?)\s+(?:dollars?|dirhams?|euros?|pounds?|rupees?|riyals?|aed|usd|eur|gbp|inr|sar)"#,
            #"(?:spent|spend|paid|pay|cost)\s+(.*?)\s+(?:on|for|at)"#,
            #"^(.*?)\s+(?:for|on|at)"#
        ]

        for patternString in amountPatterns {
            guard let regex = try? NSRegularExpression(pattern: patternString, options: .caseInsensitive) else {
                continue
            }

            let range = NSRange(location: 0, length: normalizedCommand.utf16.count)
            if let match = regex.firstMatch(in: normalizedCommand, options: [], range: range),
               let matchRange = Range(match.range(at: 1), in: normalizedCommand) {
                let amountText = String(normalizedCommand[matchRange])
                if let amount = parse(amountText) {
                    return amount
                }
            }
        }

        // Fallback: try parsing the whole command
        return parse(normalizedCommand)
    }

    /// Check if text contains number phrases
    /// - Parameter text: The text to check
    /// - Returns: True if number phrases are detected
    func containsNumberPhrase(_ text: String) -> Bool {
        let normalizedText = normalizeText(text)
        let words = normalizedText.components(separatedBy: .whitespaces)

        for word in words {
            if basicNumbers[word] != nil ||
               tensNumbers[word] != nil ||
               scaleNumbers[word] != nil {
                return true
            }
        }

        return false
    }

    // MARK: - Private Methods

    /// Normalize text for parsing (lowercase, trim, clean)
    private func normalizeText(_ text: String) -> String {
        var normalized = text.lowercased()

        // Replace hyphens with spaces for consistency
        normalized = normalized.replacingOccurrences(of: "-", with: " ")

        // Remove extra whitespace
        normalized = normalized.components(separatedBy: .whitespacesAndNewlines)
            .filter { !$0.isEmpty }
            .joined(separator: " ")

        return normalized
    }

    /// Extract numeric amount from text (e.g., "123", "1,234.56")
    private func extractNumericAmount(_ text: String) -> Double? {
        // Patterns ordered by specificity
        let patterns = [
            #"(\d+(?:,\d{3})*(?:\.\d{1,2})?)"#, // With thousand separators and decimals
            #"(\d+\.\d{1,2})"#,                   // Simple decimals
            #"(\d+)"#                              // Simple integers
        ]

        for patternString in patterns {
            guard let regex = try? NSRegularExpression(pattern: patternString, options: []) else {
                continue
            }

            let range = NSRange(location: 0, length: text.utf16.count)
            if let match = regex.firstMatch(in: text, options: [], range: range),
               let matchRange = Range(match.range(at: 1), in: text) {
                let numberString = String(text[matchRange])
                    .replacingOccurrences(of: ",", with: "")

                if let amount = Double(numberString) {
                    return amount
                }
            }
        }

        return nil
    }

    /// Parse written number from text (e.g., "two thousand five hundred")
    private func parseWrittenNumber(_ text: String) -> Double? {
        let words = text.components(separatedBy: .whitespaces)
            .filter { !$0.isEmpty && $0 != "and" } // Remove "and" connectors

        var total: Double = 0
        var current: Double = 0
        var decimalMode = false
        var decimalPlaces: [Double] = []

        for word in words {
            let cleanWord = word.lowercased()

            // Handle decimal point
            if cleanWord == "point" || cleanWord == "dot" {
                decimalMode = true
                continue
            }

            // Process decimal digits after "point"
            if decimalMode {
                if let digitValue = basicNumbers[cleanWord] {
                    decimalPlaces.append(digitValue)
                    continue
                } else {
                    // Exit decimal mode if non-digit word encountered
                    decimalMode = false
                }
            }

            // Process basic numbers (0-19)
            if let value = basicNumbers[cleanWord] {
                current += value
                continue
            }

            // Process tens (20, 30, ..., 90)
            if let value = tensNumbers[cleanWord] {
                current += value
                continue
            }

            // Process scale multipliers (hundred, thousand, million, etc.)
            if let scale = scaleNumbers[cleanWord] {
                if current == 0 {
                    current = 1 // Implicit "one" before scale (e.g., "thousand" = "one thousand")
                }

                if scale == 100 {
                    // Hundred: multiply current value
                    current *= scale
                } else {
                    // Thousand and above: multiply and add to total, reset current
                    current *= scale
                    total += current
                    current = 0
                }
                continue
            }
        }

        // Add any remaining current value to total
        total += current

        // Apply decimal places if any
        if !decimalPlaces.isEmpty {
            var decimalValue = 0.0
            var divisor = 10.0

            for digit in decimalPlaces {
                decimalValue += digit / divisor
                divisor *= 10
            }

            total += decimalValue
        }

        // Return nil if no valid number was parsed
        return total > 0 ? total : nil
    }
}

File: android/app/src/main/java/com/justspent/app/utils/NumberPhraseParser.kt
package com.justspent.app.utils

import java.math.BigDecimal
import java.util.Locale

/**
 * Comprehensive number phrase parser for voice recognition
 *
 * Supports:
 * - Basic numbers (one, two, three, etc.)
 * - Compound numbers (twenty-five, forty-two, etc.)
 * - Western scale: hundred, thousand, million, billion, trillion
 * - Indian scale: lakh, crore
 * - Fractional amounts (point five, and fifty cents, etc.)
 * - Multiple formats and natural language variations
 *
 * Examples:
 * - "two thousand" → 2000
 * - "five lakh" → 500000
 * - "two point five million" → 2500000
 * - "one hundred and fifty" → 150
 * - "twenty five thousand three hundred" → 25300
 * - "three crore" → 30000000
 *
 * @author Just Spent Team
 * @since 1.0.0
 */
object NumberPhraseParser {

    /**
     * Parse a text phrase containing number words into a BigDecimal
     *
     * @param text Input text containing number words
     * @param locale Locale for region-specific number parsing (for future use)
     * @return Parsed number as BigDecimal, or null if no valid number found
     */
    fun parse(text: String, locale: Locale = Locale.getDefault()): BigDecimal? {
        val normalizedText = normalizeText(text)

        // Try to extract numeric amount first (priority)
        extractNumericAmount(normalizedText)?.let { return it }

        // Try to parse written number phrases
        return parseWrittenNumber(normalizedText)
    }

    /**
     * Extract numeric amount from text (e.g., "2000", "2,000", "2000.50")
     */
    private fun extractNumericAmount(text: String): BigDecimal? {
        // Patterns for different numeric formats
        val patterns = listOf(
            // With thousand separators: 2,000 or 2,000.50
            Regex("""(\d{1,3}(?:,\d{3})+(?:\.\d+)?)"""),
            // Simple decimal: 2000.50
            Regex("""(\d+\.\d+)"""),
            // Simple integer: 2000
            Regex("""(\d+)""")
        )

        for (pattern in patterns) {
            val match = pattern.find(text)
            if (match != null) {
                try {
                    val amountStr = match.groupValues[1].replace(",", "")
                    return BigDecimal(amountStr)
                } catch (e: NumberFormatException) {
                    continue
                }
            }
        }

        return null
    }

    /**
     * Parse written number phrases into numeric value
     * Handles complex multi-scale numbers like "two thousand five hundred"
     */
    private fun parseWrittenNumber(text: String): BigDecimal? {
        val words = text.split(Regex("""\s+|-"""))
            .map { it.lowercase().trim() }
            .filter { it.isNotEmpty() }

        if (words.isEmpty()) return null

        var total = 0.0
        var current = 0.0
        var fractionalPart: Double? = null
        var inFractionalSection = false

        for (i in words.indices) {
            val word = words[i]

            when {
                // Handle "and" separator (common in British English)
                word == "and" -> {
                    // Just skip "and" connectors
                    continue
                }

                // Handle fractional indicators
                word == "point" -> {
                    // Next words are decimal places
                    inFractionalSection = true
                    if (current > 0) {
                        total += current
                        current = 0.0
                    }
                    continue
                }

                // Handle "cents" or "paise" (fractional currency)
                word in listOf("cents", "cent", "paise", "paisa") -> {
                    // Previous number was in cents
                    if (current > 0) {
                        fractionalPart = current / 100.0
                        current = 0.0
                    }
                    continue
                }

                // Handle basic number words (0-19)
                basicNumbers.containsKey(word) -> {
                    val value = basicNumbers[word]!!
                    if (inFractionalSection) {
                        // Building decimal places digit by digit
                        fractionalPart = (fractionalPart ?: 0.0) * 10 + value
                    } else {
                        current += value
                    }
                }

                // Handle tens (20, 30, 40, etc.)
                tensNumbers.containsKey(word) -> {
                    current += tensNumbers[word]!!
                }

                // Handle hundred
                word == "hundred" -> {
                    if (current == 0.0) current = 1.0 // "hundred" alone = 100
                    current *= 100
                }

                // Handle thousand
                word in listOf("thousand", "thousands") -> {
                    if (current == 0.0) current = 1.0 // "thousand" alone = 1000
                    current *= 1000
                    total += current
                    current = 0.0
                }

                // Handle lakh (Indian numbering system)
                word in listOf("lakh", "lakhs", "lac", "lacs") -> {
                    if (current == 0.0) current = 1.0 // "lakh" alone = 100,000
                    current *= 100000
                    total += current
                    current = 0.0
                }

                // Handle crore (Indian numbering system)
                word in listOf("crore", "crores") -> {
                    if (current == 0.0) current = 1.0 // "crore" alone = 10,000,000
                    current *= 10000000
                    total += current
                    current = 0.0
                }

                // Handle million
                word in listOf("million", "millions") -> {
                    if (current == 0.0) current = 1.0 // "million" alone = 1,000,000
                    current *= 1000000
                    total += current
                    current = 0.0
                }

                // Handle billion
                word in listOf("billion", "billions") -> {
                    if (current == 0.0) current = 1.0
                    current *= 1000000000
                    total += current
                    current = 0.0
                }

                // Handle trillion
                word in listOf("trillion", "trillions") -> {
                    if (current == 0.0) current = 1.0
                    current *= 1000000000000
                    total += current
                    current = 0.0
                }
            }
        }

        // Add any remaining current value
        total += current

        // Combine whole and fractional parts
        if (fractionalPart != null) {
            // Normalize fractional part (e.g., "point five" = 0.5, not 0.05)
            val normalizedFraction = if (inFractionalSection && fractionalPart < 1.0) {
                fractionalPart
            } else if (fractionalPart >= 1.0 && fractionalPart < 100.0) {
                // e.g., "fifty cents" = 0.50
                fractionalPart / 100.0
            } else {
                fractionalPart
            }
            total += normalizedFraction
        }

        return if (total > 0) BigDecimal.valueOf(total) else null
    }

    /**
     * Normalize input text for better parsing
     * - Convert to lowercase
     * - Remove extra spaces
     * - Handle common speech-to-text variations
     */
    private fun normalizeText(text: String): String {
        return text
            .lowercase()
            .trim()
            // Handle hyphenated numbers
            .replace("-", " ")
            // Remove common filler words
            .replace(Regex("""\b(just|spent|paid|cost|costs|pay)\b"""), "")
            // Remove currency names for cleaner parsing
            .replace(Regex("""\b(dollars?|dirhams?|euros?|pounds?|rupees?|riyals?|aed|usd|eur|gbp|inr|sar)\b"""), "")
            // Normalize multiple spaces
            .replace(Regex("""\s+"""), " ")
            .trim()
    }

    /**
     * Check if text likely contains a number phrase
     */
    fun containsNumberPhrase(text: String): Boolean {
        val normalizedText = normalizeText(text)
        val words = normalizedText.split(Regex("""\s+"""))

        // Check if any word is a number word or multiplier
        return words.any { word ->
            basicNumbers.containsKey(word) ||
            tensNumbers.containsKey(word) ||
            multipliers.contains(word)
        }
    }

    /**
     * Extract amount from complex voice commands
     * Handles patterns like "I spent two thousand dirhams on groceries"
     */
    fun extractAmountFromCommand(command: String): BigDecimal? {
        val normalizedCommand = command.lowercase()

        // Try to extract amount substring between action words and currency/category
        val amountPatterns = listOf(
            // "spent [amount] dirhams"
            Regex("""(?:spent|spend|paid|pay|cost)\s+(.*?)\s+(?:dollars?|dirhams?|euros?|pounds?|rupees?|aed|usd|eur|gbp|inr|sar)"""),
            // "spent [amount] on"
            Regex("""(?:spent|spend|paid|pay|cost)\s+(.*?)\s+(?:on|for|at)"""),
            // "[amount] for/on"
            Regex("""^(.*?)\s+(?:for|on|at)""")
        )

        for (pattern in amountPatterns) {
            val match = pattern.find(normalizedCommand)
            if (match != null) {
                val amountText = match.groupValues[1].trim()
                parse(amountText)?.let { return it }
            }
        }

        // Fallback: try to parse the entire command
        return parse(normalizedCommand)
    }

    // ===========================================
    // Number Word Mappings
    // ===========================================

    /**
     * Basic number words (0-19)
     */
    private val basicNumbers = mapOf(
        "zero" to 0.0,
        "one" to 1.0,
        "two" to 2.0,
        "three" to 3.0,
        "four" to 4.0,
        "five" to 5.0,
        "six" to 6.0,
        "seven" to 7.0,
        "eight" to 8.0,
        "nine" to 9.0,
        "ten" to 10.0,
        "eleven" to 11.0,
        "twelve" to 12.0,
        "thirteen" to 13.0,
        "fourteen" to 14.0,
        "fifteen" to 15.0,
        "sixteen" to 16.0,
        "seventeen" to 17.0,
        "eighteen" to 18.0,
        "nineteen" to 19.0,
        // Common alternatives and misspellings
        "a" to 1.0,  // "a hundred" = "one hundred"
        "an" to 1.0  // "an hour" etc.
    )

    /**
     * Tens place words (20, 30, 40, etc.)
     */
    private val tensNumbers = mapOf(
        "twenty" to 20.0,
        "thirty" to 30.0,
        "forty" to 40.0,
        "fifty" to 50.0,
        "sixty" to 60.0,
        "seventy" to 70.0,
        "eighty" to 80.0,
        "ninety" to 90.0
    )

    /**
     * Scale multipliers
     */
    private val multipliers = setOf(
        "hundred", "thousand", "thousands",
        "lakh", "lakhs", "lac", "lacs",
        "crore", "crores",
        "million", "millions",
        "billion", "billions",
        "trillion", "trillions"
    )

    // ===========================================
    // Testing & Validation Helpers
    // ===========================================

    /**
     * Validate that a number phrase parses correctly
     * Useful for testing and debugging
     */
    fun validate(phrase: String, expectedValue: Double): Boolean {
        val parsed = parse(phrase)
        return parsed != null && parsed.toDouble() == expectedValue
    }

    /**
     * Get example phrases for testing
     */
    fun getExamplePhrases(): Map<String, Double> {
        return mapOf(
            // Basic numbers
            "five" to 5.0,
            "fifteen" to 15.0,
            "twenty" to 20.0,
            "twenty five" to 25.0,
            "ninety nine" to 99.0,

            // Hundreds
            "one hundred" to 100.0,
            "two hundred" to 200.0,
            "five hundred and fifty" to 550.0,
            "nine hundred ninety nine" to 999.0,

            // Thousands
            "one thousand" to 1000.0,
            "two thousand" to 2000.0,
            "ten thousand" to 10000.0,
            "twenty five thousand" to 25000.0,
            "one hundred thousand" to 100000.0,
            "two thousand five hundred" to 2500.0,

            // Lakhs (Indian system)
            "one lakh" to 100000.0,
            "five lakh" to 500000.0,
            "ten lakh" to 1000000.0,
            "twenty five lakh" to 2500000.0,

            // Crores (Indian system)
            "one crore" to 10000000.0,
            "two crore" to 20000000.0,
            "five crore" to 50000000.0,

            // Millions
            "one million" to 1000000.0,
            "two million" to 2000000.0,
            "five million" to 5000000.0,
            "ten million" to 10000000.0,

            // Billions
            "one billion" to 1000000000.0,
            "two billion" to 2000000000.0,

            // Complex combinations
            "two thousand five hundred and fifty" to 2550.0,
            "one million two hundred thousand" to 1200000.0,
            "five lakh fifty thousand" to 550000.0,

            // Decimals
            "five point five" to 5.5,
            "two point five million" to 2500000.0,
            "one point two five" to 1.25,

            // With "and" connector
            "one hundred and twenty" to 120.0,
            "two thousand and five" to 2005.0
        )
    }
}
